src/Integration/MefServices/ProjectsRuleSetProvider.cs

```diff
@@ -0,0 +1,355 @@
+        }
+
+        private async void OnActiveSolutionChanged(object sender, ActiveSolutionChangedEventArgs e)
+        {
+            ClearCache();
+
+            if (e.IsSolutionOpen)
+            {
+                await BuildCacheAsync();
+            }
+        }
+
+        public bool HasRuleSetWithSonarAnalyzerRules(string projectFilePath) =>
+            projectPathToCachedData.ContainsKey(projectFilePath) && projectPathToCachedData[projectFilePath].HasAnySonarRule;
+
+        private void ClearCache()
+        {
+            filesToTrack.Values.ToList().ForEach(x => x.UpdatedOnDisk -= OnUpdatedOnDisk);
+            filesToTrack.Clear();
+
+            projectPathToCachedData.Clear();
+        }
+
+        private async Task BuildCacheAsync()
+        {
+            while (!projectSystemHelper.IsSolutionFullyOpened())
+            {
+                await Task.Delay(2000); // REVIEW: Is 2s a good enough time?
+            }
+
+            projectSystemHelper.GetSolutionProjects()
+                .ToList()
+                .ForEach(p => projectPathToCachedData.Add(p.FullName, ProcessProjectAndCollectFilesToTrack(p)));
+
+            // Subscribe to file changes for the projects and the rulesets
+            filesToTrack.Values.ToList().ForEach(x => x.UpdatedOnDisk += OnUpdatedOnDisk);
+
+            if (this.configurationProvider.GetConfiguration().Mode == SonarLintMode.Connected)
+            {
+                var projectsWithSonarRuleSet = projectPathToCachedData.Where(x => x.Value.HasAnySonarRule)
+                    .Select(x => Path.GetFileNameWithoutExtension(x.Key))
+                    .ToList();
+                if (projectsWithSonarRuleSet.Count > 0)
+                {
+                    this.logger.WriteLine(Strings.WarningProjectsWithSonarRulesWhileNewConnected,
+                        string.Join(", ", projectsWithSonarRuleSet));
+                }
+            }
+        }
+
+        private async void OnUpdatedOnDisk(object sender, EventArgs e)
+        {
+            var fileChangeTracker = sender as FileChangeTracker;
+            if (fileChangeTracker == null)
+            {
+                return;
+            }
+
+            if (!fileChangeTracker.FilePath.EndsWith(".ruleset"))
+            {
+                var cachedProjectRuleSetPath = projectPathToCachedData[fileChangeTracker.FilePath].RuleSetPath;
+
+                var project = projectSystemHelper.GetSolutionProjects().FirstOrDefault(p =>
+                    p.FullName.Equals(fileChangeTracker.FilePath, StringComparison.OrdinalIgnoreCase));
+                if (project != null)
+                {
+                    var projectRuleSetPath = projectSystemHelper.GetProjectProperty(project, Constants.CodeAnalysisRuleSetPropertyKey);
+                    var projectDirectoryFullPath = new FileInfo(project.FullName).Directory.FullName;
+                    var projectRuleSetFullPath = GetFullPath(projectRuleSetPath, projectDirectoryFullPath);
```

Comment:
181-183 - candidate for a function? I think this logic occurs more than once.

Reply:
It's indeed called twice but creating a method returning 3 parameters is not ideal. Creating a class with 3 properties readonly for that is overkill in c#.